### PR TITLE
Add doctor validation for host requered for some components to work.

### PR DESCRIPTION
**Pull request by Maksimka101**
## Description

For example, checking the availability of pub.dev.

## Related Issues

https://github.com/flutter/flutter/issues/49403

## Tests

I added the following tests:
- tests for host validator to doctor_test.dart

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ x] I signed the [CLA].
- [ x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ x] I updated/added relevant documentation (doc comments with `///`).
- [ x] All existing and new tests are passing.
- [ x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Its kind of hard to understand the change with the formatting diff. Could you revert that?

The tool  codebase does not use dartfmt, so everything needs to be manually formatted

File: packages/flutter_tools/lib/src/host_validator.dart [hunk 1]
```diff
@@ -0,0 +1,90 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:http/http.dart' as http;
+
+import 'doctor.dart';
+import 'globals.dart';
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/host_validator.dart:8]: This line needs to be `import 'globals.dart' as globals;`

File: packages/flutter_tools/lib/src/host_validator.dart [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  @override
+  Future<ValidationResult> validate() async {
+    final List<ValidationMessage> messages = <ValidationMessage>[];
+    final Iterable<Future<_HostValidationResult>> availabilityResultFuture =
+        _requiredHosts.map(_checkHostAvailability);
+
+    final List<_HostValidationResult> availabilityResult =
+        (await Future.wait(availabilityResultFuture)).toList();
+
+    /// if every hosts are reachable
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/host_validator.dart:56]: Use `//` for comments

File: packages/flutter_tools/lib/src/host_validator.dart [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+  /// Hosts which are required for flutter
+  /// For flutter build for example
+  List<String> get _requiredHosts {
+    final List<String> hosts = <String>[
+      'https://pub.dev/',
+      'https://cloud.google.com/',
+      'https://maven.google.com/',
+    ];
+    if (platform.isMacOS) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/host_validator.dart:27]: Instead of using the values in globals.dart, inject these in the constructor like:

```
HostValidator({
  @required Platform platform,
}) : _platform = platform;

final Platform _platform;
```

File: packages/flutter_tools/lib/src/host_validator.dart [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+  @override
+  String get slowWarning => 'Host availability check is taking a long time...';
+
+  http.Client httpClient = http.Client();
+
+  /// Hosts which are required for flutter
+  /// For flutter build for example
+  List<String> get _requiredHosts {
+    final List<String> hosts = <String>[
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/host_validator.dart:22]: hoist these to a const list outside of the validator

File: packages/flutter_tools/lib/src/host_validator.dart [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import 'globals.dart';
+
+/// Verify that the required hosts ( they are stored at [_requiredHosts] ) are reachable
+class HostValidator extends DoctorValidator {
+  HostValidator() : super('Host availability');
+
+  @override
+  String get slowWarning => 'Host availability check is taking a long time...';
+
+  http.Client httpClient = http.Client();
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/host_validator.dart:17]: The http.Client should be injected in the constructor so that we can avoid making network requests during the test

File: packages/flutter_tools/test/commands.shard/hermetic/doctor_test.dart [hunk 6]
```diff
@@ -221,18 +243,69 @@ void main() {
     });
   });
 
+  group('host validator', () {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/doctor_test.dart:246]: Place these tests in a host_validator_test.dart file instead of adding to the doctor_test.dart

File: packages/flutter_tools/test/commands.shard/hermetic/doctor_test.dart [hunk 7]
```diff
@@ -35,9 +35,9 @@ import '../../src/context.dart';
 import '../../src/testbed.dart';
 
 final Generator _kNoColorOutputPlatform = () => FakePlatform(
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/doctor_test.dart:37]: Please revert unrelated changes

**Comment by jonahwilliams** [comment]
It looks like the commit you picked up might have a broken macOS build, you could pull in master to fix

**Comment by zanderso** [comment]
@Maksimka101 Since this PR looks inactive, I'm going to close it, but feel free to re-open if you'd like to continue working on it.
